sklearn/utils/_test_common/instance_generator.py

```diff
@@ -542,6 +542,17 @@
         ]
     },
     MDS: {"check_dict_unchanged": dict(max_iter=5, n_components=1, n_init=2)},
+    MLPClassifier: {
+        "check_sample_weight_equivalence": [
+            dict(solver="lbfgs"),
+            dict(solver="lbfgs", early_stopping=True, n_iter_no_change=1, tol=1e-2),
+        ]
+    },
+    MLPRegressor: {
+        "check_sample_weight_equivalence": [
+            dict(solver="sgd", tol=1e-2, random_state=42),
+        ]
+    },
```

Comment:
We could indeed have an estimator-specific estimator with a fixed random seed to check that it's doing something reasonable on a synthetic dataset crafted specifically to lead to very different predictions if the weights were not handled properly:


```python
import numpy as np
from sklearn.neural_network import MLPClassifier

X = np.asarray([0, 0, 0, 1, 1, 1, 1]).reshape(-1, 1)
y = np.asarray([0, 1, 1, 2, 2, 2, 1])
sw = np.asarray([2, 0, 0, 0.1, 0.1, 0.1, 3])

global_random_seed = None  # TODO: set me when adapting this to pytest
solver = "sgd"
clf = MLPClassifier(
    random_state=global_random_seed, solver=solver, max_iter=10_000
).fit(X, y, sample_weight=sw)
np.testing.assert_array_equal(clf.predict([[0], [1]]), [0, 1])
```

I would expect this to pass for all solvers and for any value of `global_random_seed`.

Reply:
@glemaitre Do we really need [that](https://github.com/scikit-learn/scikit-learn/pull/30155#pullrequestreview-2420470861)? You can read the code and convince yourself that sample weights are passed. I have the impression that we get neurotic if we go for that kind of rigor.

Reply:
For me the benefit is twofold:

- We had in the past in multiple estimators this bug where we don't pass the sample weight the evaluation internal metric (for early-stopping or internal CV). So there is not reason that this issue happen in the future if we don't have a non-regression test.
- IMO, the test also serve as documentation: e.g. I could find about it when going through the test suite of the histogram gradient boosting.

So even though we might never need to touch this code and it might seem an overkill, I still find it useful.

Reply:
I hesitate to commit this.

Reply:
I would rather test values of `best_validation_score_` than the monkeypatch solution.

Reply:
I've to stop for today.

Reply:
> I would rather test values of best_validation_score_ than the monkeypatch solution.

This is a viable alternative as well. I might get some time to make a small test tonight inside this direction.